libs/metrics-core/src/Data/Metrics.hs

```diff
@@ -2,214 +2,281 @@
     , gaugeSet
     , gaugeValue
 
-    , labelGet
-    , labelGet'
-    , labelSet
-    , labelValue
+    -- * Histograms
+    -- ** Types
+    , HistogramInfo
+    , Buckets
+    , Bucket
+
+    -- ** Describing Histograms
+    , linearHistogram
+    , customHistogram
+    , deprecatedRequestDurationHistogram
 
-    , bucketsGet
-    , bucketsGet'
-    , bucketsIncr
+    -- ** Manipulating Histograms
+    , histoGet
+    , histoSubmit
+    , histoValue
+    , histoTimeAction
 
+    -- * Helper functions
+    , path
+    , metrics
     , render
     ) where
 
 import Imports hiding (lookup, union)
 import Data.Aeson
-import Data.Atomics.Counter (AtomicCounter)
 import Data.Hashable
-import Data.Metrics.Buckets (Buckets)
 
-import qualified Data.Atomics.Counter as Atomic
 import qualified Data.Text            as T
-import qualified Data.HashMap.Strict  as Map
-import qualified Data.Metrics.Buckets as Buckets
+import qualified Data.HashMap.Strict  as HM
+import qualified Data.Map.Strict      as M
 import qualified Data.Metrics.GC      as GC
-
-newtype Path = Path { _path :: Text } deriving (Eq, Hashable, Semigroup, Monoid)
-
+import qualified Prometheus as P
+
+-- | Internal Counter type
+type Counter = P.Counter
+-- | Internal Gauge type
+type Gauge = P.Gauge
+-- | Internal Histogram type
+type Histogram = P.Histogram
+
+-- | Represents a descriptive metric path or name.
+--
+-- NOTE: Until all metrics are fully migrated to Prometheus this should be a valid
+-- name according to collectd; e.g. @net.resources./teams/invitations/info@
+-- All names are converted into valid prometheus names when needed via 'toInfo'
+newtype Path =
+    Path
+        { _path :: Text
+        }
+    deriving (Eq, Show, Hashable, Semigroup, Monoid)
+
+-- | Create a path
 path :: Text -> Path
 path = Path
 
-newtype Counter = Counter AtomicCounter
-newtype Gauge   = Gauge   AtomicCounter
-newtype Label   = Label   (IORef Text)
-
-data Metrics = Metrics
-    { counters :: IORef (HashMap Path Counter)
-    , gauges   :: IORef (HashMap Path Gauge)
-    , labels   :: IORef (HashMap Path Label)
-    , buckets  :: IORef (HashMap Path Buckets)
-    }
+-- | Opaque storage of metrics
+data Metrics =
+    Metrics
+        { counters   :: IORef (HashMap Path Counter)
+        , gauges     :: IORef (HashMap Path Gauge)
+        , histograms :: IORef (HashMap Path Histogram)
+        }
 
+-- Initialize an empty set of metrics
 metrics :: MonadIO m => m Metrics
 metrics = liftIO $ Metrics
-    <$> newIORef Map.empty
-    <*> newIORef Map.empty
-    <*> newIORef Map.empty
-    <*> newIORef Map.empty
+    <$> newIORef HM.empty
+    <*> newIORef HM.empty
+    <*> newIORef HM.empty
+
+-- | Converts a CollectD style 'path' to a Metric name usable by prometheus
+--   This is to provide back compatibility with the previous collect-d metric names
+--   which often had paths and dot-separated names.
+--
+-- Currently just replaces all "/" and "." with "_" and lowercases the result
```

Comment:
If this could be done, that'd be nice, as it would make it easier for someone to understand some of these metrics if they had idiomatic names and a description. Migrating from un-idiomatic no-description prometheus metrics to idiomatically named ones won't be made easy if we wish to keep backwards compatibility. I'm thinking it'd be nice to only adapt grafana dashboards once, not twice.

Reply:
I took over a few of the names; and honestly most seem to be very good prometheus names already, with the exception of perhaps adding `_total` to the end of a few of them; descriptions would also be nice, but that's something that doesn't break any interfaces so it can easily be adjusted at a later date; this PR is about being backwards compatible; we can focus on improving the state of things in a future one 😄